Prop definition-driven styles with support for arbitrary responsive values

vladmoroz opened the pull request:
- General
  - Add the following props to all layout components:
    - `minWidth`, `maxWidth`
    - `minHeight`, `maxHeight`
    - `flexBasis`, `flexShrink`, `flexGrow`
    - `gridColumn`, `gridColumnStart`, `gridColumnEnd`
    - `gridRow`, `gridRowStart`, `gridRowEnd`
  - Rework all layout props to allow arbitrary CSS values, including when used with the responsive object syntax. Props that support arbitrary values include:
    - `width`, `minWidth`, `maxWidth`
    - `height`, `minHeight`, `maxHeight`
    - `m`, `mx`, `my`, `mt`, `mr`, `mb`, `ml`
    - `p`, `px`, `py`, `pt`, `pr`, `pb`, `pl`
    - `inset`, `top`, `right`, `bottom`, `left`
    - `gap`, `gapX`, `gapY`
    - `flexBasis`, `flexShrink`, `flexGrow`
    - `gridColumn`, `gridColumnStart`, `gridColumnEnd`
    - `gridRow`, `gridRowStart`, `gridRowEnd`
  - Deprecate `shrink` and `grow` props in favour of `flexShrink` and `flexGrow`. The older `shrink` and `grow` props will be removed in the next major release.
  - Update the type signature of the layout props so that code editor suggestions use just space scale values when possible. CSS keywords and other values such as `"auto"` or `"100vw"` are still available as manual string values.
- `Container`, `Section`
  - Change the incorrect `display="block"` value to `display="initial"`
- `Checkbox`, `RadioGroup`, `Switch`
  - Rework the internal HTML structure and styles. These components now render fewer HTML nodes and forward all props to the topmost node. Make sure that your code works as expected if you were relying on any of the implementation quirks to override styles or behaviour.
- `Flex`
  - Add `gapX` and `gapY` props
- `Table`
  - Add new `layout` prop to control the `table-layout` style property
  - Align `width` prop type signature and implementation on the `TableCell` part with the reworked `width` prop on the layout components
  - Add `minWidth` and `maxWidth` props to the `TableCell` part
- `TextArea`
  - Add `radius` prop

vladmoroz commented on apps/playground/package.json:
```diff
@@ -24,6 +24,6 @@
     "@types/react-dom": "^18.0.7",
     "eslint-config-custom": "*",
-    "typescript": "^4.5.3"
+    "typescript": "^5.3.3"
```
Upgraded it while I was in the weeds of TS. Now that I'm through it, don’t think we absolutely _have to_, but decided to keep it anyway.

benoitgrelard commented on packages/radix-ui-themes/src/helpers/responsive.ts:
```diff
@@ -0,0 +1,181 @@
+  const classNames: string[] = [];
+
+  if (!value) {
+    return '';
```
nit: That would feel more accurate to the return signature.
```suggestion
    return undefined;
```

benoitgrelard commented on packages/radix-ui-themes/src/components/alert-dialog.tsx:
```diff
@@ -35,26 +35,18 @@ interface AlertDialogContentProps
       <AlertDialogPrimitive.Portal container={container} forceMount={forceMount}>
         <Theme asChild>
-          <AlertDialogPrimitive.Overlay className="rt-DialogOverlay rt-AlertDialogOverlay">
+          <AlertDialogPrimitive.Overlay className="rt-AlertDialogOverlay rt-DialogOverlay">
```
There are a few files where classes have been reordered when there was consistency in the codebase so far:
- `AlertDialog`
- `ContextMenu`
- `DropdownMenu`
- `Link`
- `RadioGroup`

benoitgrelard commented on packages/radix-ui-themes/src/helpers/responsive.ts:
```diff
@@ -0,0 +1,181 @@
+
+// Split comma-separated custom properties. This is an escape hatch for
+// when you need to generate multiple custom properties with the same value.
+function getCustomProperties(str: `--${string}`) {
```
Should we be more explicit and accept this?
```ts
`--${string}` | `--${string}`[]
```

so instead of:

```ts
customProperty: '--margin-left, --margin-right',
```

we pass:

```ts
customProperty: ['--margin-left', '--margin-right'],
```

benoitgrelard replied:
I feel like we should yes, because that's clearly the intent here that it can have single or multiple custom properties, so using string with a separator to split on feels like a hack we're shielding away from the compiler.

vladmoroz replied:
Updated to
```ts
customProperties: `--${string}`[]
```

The middle ground was fancy-ish, but not fancy enough to keep

benoitgrelard commented on packages/radix-ui-themes/src/helpers/extract-props.ts:
```diff
@@ -0,0 +1,94 @@
+      }
+
+      if (propDef.type === 'boolean' && value) {
+        // TODO handle responsive boolean props
```
Is that because we don't have a use case for it yet?

vladmoroz commented on packages/radix-ui-themes/src/styles/utilities/margin.css:
```diff
@@ -211,676 +195,885 @@
+      var(--margin-bottom-override, var(--margin-bottom))
+      var(--margin-left-override, var(--margin-left));
+  }
+}
```
We could find a way to do it, not sure it's worth it though

benoitgrelard replied:
Yeah don't think we need to

vladmoroz commented on packages/radix-ui-themes/src/styles/utilities/min-width.css:
```diff
@@ -0,0 +1,66 @@
+  .xl\:rt-r-min-w {
+    min-width: var(--min-width-xl);
+  }
+}
```
We could do it with something like `@breakpoints var(--min-width)`, where the at-rule would have optional arguments to let it know which variables need a breakpoints suffix. Don't think it's worth it yet, the copy-pasta bit hasn't been too annoying so far.

benoitgrelard commented on packages/radix-ui-themes/src/components/select.tsx:
```diff
@@ -39,31 +39,21 @@ interface SelectTriggerProps
-    const { size } = React.useContext(SelectContext);
+    const context = React.useContext(SelectContext);
+    const { className, color, radius, placeholder, ...triggerProps } = extractProps(
+      { size: context?.size, ...props },
+      { size: selectRootPropDefs.size },
```
Ohhhhhhh I finally understand those now!!!

- The first  `{ size: context?.size, ...props }` is the new "props"
- The second `{ size: selectRootPropDefs.size }` is some extra prop defs we want to match

This was super unclear to me when I read it, I feel we should add some comments on those as they stand out.
Something like:

```tsx
{ size: context?.size, ...props }, // we pretend there's also a `size` prop
{ size: selectRootPropDefs.size }, // we add a matching prop def to extract that `size` prop
```

vladmoroz replied:
Does this feel good?

<img width="724" alt="image" src="https://github.com/radix-ui/themes/assets/8441036/ef8195aa-b628-4e88-96fa-9dab5075e9bc">

vladmoroz replied:
Pushed 98b0ed932e1d05aff0e246ee7e342524ff429976

benoitgrelard replied:
Yes that's great, if we can do that on each of these stand out use case that'd be 👌

vladmoroz commented on packages/radix-ui-themes/src/styles/utilities/padding.css:
```diff
@@ -1,331 +1,505 @@
   }
-
+  .rt-r-px {
+    /* Keep this selector here to maintain the source order */
```
OK as I started to reply to this comment and wanted to copy-paste an example of the resulting CSS after the transformations, I realized that one had a significant flaw and reworked some stuff to fix it. Good question 🙂 

Anyway, in the resulting CSS, we want to have all related `.xs:`, `.md:`, `.lg:`, etc rules grouped together like this:

```css
.rt-r-inset { ... }
.rt-r-inset-0 { ... }
.rt-r-inset-1 { ... }
...

@media (--xs) {
  .xs:rt-r-inset { ... }
  .xs:rt-r-inset-0 { ... }
  .xs:rt-r-inset-1 { ... }
  ...
}

@media (--md) {
  .md:rt-r-inset { ... }
  .md:rt-r-inset-0 { ... }
  .md:rt-r-inset-1 { ... }
  ...
}

...
```

Adding an empty rule inside @breakpoints allows us to take advantage of how our PostCSS setup combines the duplicate selectors, so I'm deliberately placing the matching selectors in the right spot while they are defined later below @breakpoints.

Otherwise, without placing the rule in the right spot first, the result would have been this:

```css
.rt-r-inset-0 { ... }
.rt-r-inset-1 { ... }
...

@media (--xs) {
  .xs:rt-r-inset-0 { ... }
  .xs:rt-r-inset-1 { ... }
  ...
}

@media (--md) {
  .md:rt-r-inset-0 { ... }
  .md:rt-r-inset-1 { ... }
  ...
}
...

.rt-r-inset { ... }

@media (--xs) {
  .xs:rt-r-inset { ... }
}

@media (--md) {
  .md:rt-r-inset { ... }
}
...
```

This would have made it so that rules like `rt-r-inset` or `xs:rt-r-inset` would override the preceding hardcoded values like `lg:rt-r-inset-0`

vladmoroz commented:
Some changes I've made when working out how to nail the order (https://github.com/radix-ui/themes/pull/235#discussion_r1453434075), commit https://github.com/radix-ui/themes/pull/235/commits/3de5ead343d641ed4d933086a572c022a42feaf6

- Updated the PostCSS plugin so that the external plugins can process the original rule (something to do with cloning vs using the original rule)
- Removed redundant `@media all` declarations (not strictly needed, but something I was exploring as part of the potential fix, which didn't help, but decided it's a good change to keep)
- Simplified margin styles
- Added a plugin that cleans up the whitespace mess in the rules after all the transformations

vladmoroz commented:
Also upgraded our PostCSS and stylelint versions, no effective changes in the resulting CSS though
